### PR TITLE
feat: Log Reduction + Level Logging

### PR DESCRIPTION
**Pull request by davidterpay**

### CODE REVIEW

File: oracle/config/oracle.go [hunk 1]
```diff
@@ -25,6 +25,8 @@ type OracleConfig struct {
 
 	// Production specifies whether the oracle is running in production mode. This is used to
 	// determine whether the oracle should be run in debug mode or not.
+	//
+	// TODO:  Deprecate this field in favor of a more robust configuration system.
```
Comment (davidterpay) [inline oracle/config/oracle.go]: Im assuming we dont want to deprecate this till a major release?

Comment (aljo242) [reply oracle/config/oracle.go]: I think we can deprecate now and then remove at a later time.

can add a godoc like [this](https://zchee.github.io/golang-wiki/Deprecated/)

**Review by aljo242 — APPROVED** [review]
lgtm - i think we should just deprecate the `Production` field now

File: cmd/slinky/main.go [hunk 2]
```diff
@@ -77,6 +80,20 @@ func init() {
...
+	rootCmd.Flags().StringVarP(
+		&logLevel,
+		"log-level",
+		"",
+		"info",
+		"Log level (debug, info, warn, error, dpanic, panic, fatal).",
+	)
+	rootCmd.Flags().StringVarP(
+		&writeLogsTo,
+		"write-logs-to",
```
Comment (Eric-Warehime) [inline cmd/slinky/main.go]: Can we make this `log-file`, that way we have
`--log-level` and `--log-file`.

File: pkg/log/zap.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	if config.WriteTo != "" {
+		outputPaths = append(outputPaths, config.WriteTo)
+	}
+
+	// initialFields are the fields that are added to every log message.
+	initialFields := map[string]interface{}{
+		"pid": os.Getpid(),
+	}
+	if config.WriteTo != "" {
+		initialFields["writing_logs_to"] = config.WriteTo
```
Comment (Eric-Warehime) [inline pkg/log/zap.go]: This means every log will have a property like the following:
`"writing_logs_to": "<the file you're currently looking at>"` right?

**Review by Eric-Warehime — APPROVED** [review]
LGTM--small comments

File: pkg/log/zap.go (hunk 3 above)
Comment (davidterpay) [reply pkg/log/zap.go]: now that i think about it we probably should have it be a part of every log.

Comment (davidterpay) [reply pkg/log/zap.go]: removed it so it is not part of every log

File: cmd/slinky/main.go [hunk 4]
```diff
@@ -77,6 +80,20 @@ func init() {
...
 		"Port for the pprof server to listen on.",
 	)
+	rootCmd.Flags().StringVarP(
+		&logLevel,
+		"log-level",
+		"",
+		"info",
+		"Log level (debug, info, warn, error, dpanic, panic, fatal).",
+	)
+	rootCmd.Flags().StringVarP(
```
Comment (g3ndo9kar9) [inline cmd/slinky/main.go:90]: Why do we need this? @Eric-Warehime is this standard practice? My intuition is that all infra deployments assume logs are piped to stdout, and will redirect wherever they want from there (i.e in solve)

Comment (g3ndo9kar9) [reply cmd/slinky/main.go:90]: see [here](https://github.com/skip-mev/solve/blob/main/logging/logging.go#L21)

File: cmd/slinky/main.go [hunk 5]
```diff
@@ -109,18 +126,12 @@ func runOracle() error {
...
-		if err != nil {
-			return fmt.Errorf("failed to create logger: %s", err.Error())
-		}
-	}
+	// Build logger.
+	logger := log.NewLogger(log.Config{
+		LogLevel: logLevel,
+		WriteTo:  writeLogsTo,
+	})
+	defer logger.Sync()
```
Comment (g3ndo9kar9) [inline cmd/slinky/main.go:134]: What does this do?

File: pkg/log/zap.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	case "error":
+		logLevel = zapcore.ErrorLevel
+	case "dpanic":
+		logLevel = zapcore.DPanicLevel
+	case "panic":
+		logLevel = zapcore.PanicLevel
+	case "fatal":
+		logLevel = zapcore.FatalLevel
+	default:
+		logLevel = zapcore.InfoLevel
```
Comment (g3ndo9kar9) [inline pkg/log/zap.go:60]: Can't we get all of this directly from [zap](https://arc.net/l/quote/yccvklvf)

File: pkg/log/zap.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	// outputPaths is the list of paths to write logs to. By default, logs are
+	// written to stderr.
+	outputPaths := []string{"stderr"}
+	if config.WriteTo != "" {
+		outputPaths = append(outputPaths, config.WriteTo)
+	}
+
+	// initialFields are the fields that are added to every log message.
+	initialFields := map[string]interface{}{
+		"pid": os.Getpid(),
```
Comment (g3ndo9kar9) [inline pkg/log/zap.go:36]: Why do we need this? Aren't all logs logged from the same process anyway?

File: pkg/log/zap.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+		logLevel = zapcore.PanicLevel
+	case "fatal":
+		logLevel = zapcore.FatalLevel
+	default:
+		logLevel = zapcore.InfoLevel
+	}
+
+	zapConfig := zap.Config{
+		Level:             zap.NewAtomicLevelAt(logLevel),
+		Development:       dev,
```
Comment (g3ndo9kar9) [inline pkg/log/zap.go:65]: What does this field mean?

**Review by g3ndo9kar9 — COMMENTED** [review]
comments

File: cmd/slinky/main.go (hunk 4 above)
Comment (davidterpay) [reply cmd/slinky/main.go:90]: more of a utility thing than anything else

File: cmd/slinky/main.go (hunk 5 above)
Comment (davidterpay) [reply cmd/slinky/main.go:134]: blocks on process closing till all logs are written

File: pkg/log/zap.go (hunk 7 above)
Comment (davidterpay) [reply pkg/log/zap.go:36]: utility field more than anything else

Comment (davidterpay) [reply pkg/log/zap.go:36]: otherwise you have to lsof

File: pkg/log/zap.go (hunk 8 above)
Comment (davidterpay) [reply pkg/log/zap.go:65]: prints more in depth stack traces in the case where it is dev

**Pull request merged**